File: trees_graphs/deepest_leaves_sum.js
// Given the root of a binary tree, return the sum of values of its deepest leaves.

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var deepestLeavesSum = function(root) {
    if (root == null) return [];

    let queue = [root];
    let maxVal = 0;
    while(queue.length) {
        let currLength = queue.length;
        maxVal = 0;
        // We could use nextQueue implementation so we don't have to perform shift since is more expensive
        // let nextQueue = [];

        for (let i = 0; i < currLength; i++) {
            let node = queue.shift();
            // let node = queue[i]; 

            maxVal = maxVal + node.val;

            if(node.left) {
                queue.push(node.left);
                // nextQueue.push(node.left);
            }

            if(node.right) {
                queue.push(node.right);
                // nextQueue.push(node.right);
            }
        }
        // queue = nextQueue;
    }
    return maxVal;
};

File: trees_graphs/deepest-leaves-sum.java
import java.util.ArrayDeque;
import java.util.LinkedList;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public int deepestLeavesSum(TreeNode root) {
        
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        int ans = 0;
        while (!queue.isEmpty()) {
            int rowSum = 0;
            int size = queue.size();

            for (int i = 0; i < size; i++) {
                TreeNode curr = queue.remove();

                rowSum += curr.val;

                if(curr.left != null) {
                    queue.add(curr.left);   
                }
                if(curr.right != null) {
                    queue.add(curr.right);   
                }
            }
            ans = rowSum;
        }
        return ans;
    }
}